### PR TITLE
Fix `Call to a member function getId() on int`

### PR DESCRIPTION
**Pull request by convenient**
```
Call to a member function getId() on int
Exception in vendor/magento/module-customer/Block/Adminhtml/Edit/Tab/Newsletter.php:421
```

The definition for `getCurrentCustomerId()` always casts it as an `(int)` so doing `$customerId = (int)$customer->getId();` on an int is destined to fail.

https://github.com/magento/magento2/blob/6729b6e01368248abc33300208eb292c95050203/app/code/Magento/Customer/Block/Adminhtml/Edit/Tab/Newsletter.php#L288-L291

We had some custom code which called `getStatusChangedDate` when viewing a customer within the admin panel, which highlighted this call does not work.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by convenient** [comment]
@magento run all tests

**Comment by convenient** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Review by hws47a — CHANGES_REQUESTED** [review]
Hi @convenient!

Thank you for your contribution. 

The change looks good but unfortunately it cannot be approved until it meet [Definition of Done criteria](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/#definition-of-done-1)
> All changes must be covered by automated tests if possible.

You can write a simple test in `app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php` similar to `testInitFormWithCustomerFormData` (but much simpler) to cover `getStatusChangedDate()` method.

Let me know if you need any help.

**Comment by convenient** [comment]
@magento run all tests

**Comment by convenient** [comment]
@magento run static tests, unit tests

**Comment by convenient** [comment]
@magento run static tests

**Comment by convenient** [comment]
@magento run all tests

**Comment by convenient** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by convenient** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by convenient** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by convenient** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by convenient** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
Hi @convenient,

Thank you for the contribution!

It seems that the new automation test has been added. Hence requesting for the review.

Thank you!

**Review by engcom-Hotel — APPROVED** [review]
The changes look good. Hence approving the PR.

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
:heavy_check_mark: QA Passed
Fix Call to a member function getId() on int

**Manual testing scenario:**
1. Create custom module and call `getStatusChangedDate` https://github.com/magento/magento2/blob/6729b6e01368248abc33300208eb292c95050203/app/code/Magento/Customer/Block/Adminhtml/Edit/Tab/Newsletter.php#L415-L424

**Before:** :heavy_multiplication_x: 
Screenshot 2023-04-19 at 1 58 22 PM

Screenshot 2023-04-19 at 1 08 50 PM

**After:** :heavy_check_mark:  
No error is seen.
Screenshot 2023-04-19 at 1 51 57 PM

There are build failures Hence moving to `Extended Testing`

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B,Functional Tests CE,Functional Tests EE

File: app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php [hunk 1]
```diff
@@ -160,6 +158,39 @@ public function testInitFormCanNotShowTab()
...
+            ->setMethods(['loadByCustomer', 'getChangeStatusAt', 'isSubscribed', 'getData'])
+            ->getMock();
+
+        $subscriberMock->method('loadByCustomer')->with($customerId, $websiteId)->willReturnSelf();
+        $subscriberMock->method('getChangeStatusAt')->willReturn('');
+        $subscriberMock->method('isSubscribed')->willReturn($isSubscribed);
+        $subscriberMock->method('getData')->willReturn([]);
+        $this->subscriberFactoryMock->expects($this->once())->method('create')->willReturn($subscriberMock);
+
+        $this->assertEquals('', $this->model->getStatusChangedDate());
```
Comment (hws47a) [inline app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php]: It seems like the test only covers the case where `getChangeStatusAt` returns `''` and therefore `getStatusChangedDate()===''`. 
Could you please also cover a case with not-empty data?

**Comment by engcom-Echo** [comment]
Functional Tests B2B failure are different on last two run on same commit.
Run1
36769-B2B

Run2
36769-B2B2

Hence moving to `Merge In Progress`

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php [hunk 2]
```diff
@@ -161,6 +168,55 @@ public function testInitFormCanNotShowTab()
...
+     * Data provider for testGetSubscriberStatusChangedDate
+     *
+     * @return array
+     */
+    public function getChangeStatusAtDataProvider()
+    {
+        return
+            [
+                ['',''],
+                ['Nov 22, 2023, 1:00:00 AM','Nov 22, 2023, 1:00:00 AM']
```
Comment (hws47a) [inline app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php]: To make sure that formatDate is run, it's better to make expected value different to `getChangeStatusAt`

File: app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php [hunk 3]
```diff
@@ -161,6 +168,55 @@ public function testInitFormCanNotShowTab()
...
+        $customer->method('getWebsiteId')->willReturn($websiteId);
+        $customer->method('getStoreId')->willReturn($storeId);
+        $customer->method('getId')->willReturn($customerId);
+        $this->customerRepository->method('getById')->with($customerId)->willReturn($customer);
+
+        $subscriberMock = $this->getMockBuilder(Subscriber::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['loadByCustomer', 'getChangeStatusAt', 'isSubscribed', 'getData'])
+            ->getMock();
+        $this->localeDateMock->method('formatDateTime')->willReturn($statusDate);
```
Comment (hws47a) [inline app/code/Magento/Customer/Test/Unit/Block/Adminhtml/Edit/Tab/NewsletterTest.php]: And ensure that the `getChangeStatusAt` value is provided to the formatDate
```suggestion
        $this->localeDateMock->method('formatDateTime')->with($statusDate)->willReturn($statusDate);
```

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by hws47a — APPROVED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by Indraniks** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run WebAPI Tests,Functional Tests CE

**Pull request merged**